src/pydvl/influence/frameworks/torch_differentiable.py

```diff
@@ -19,19 +20,201 @@
 logger = logging.getLogger(__name__)
 
 
-def flatten_gradient(grad):
+def flatten_gradient(grad) -> torch.Tensor:
     """
     Simple function to flatten a pyTorch gradient for use in subsequent calculation
     """
     return torch.cat([el.reshape(-1) for el in grad])
 
 
-def solve_linear(matrix: torch.Tensor, b: torch.Tensor):
-    """Computes the solution of a square system of linear equations"""
-    return torch.linalg.solve(matrix, b)
+def solve_linear(
+    model: TwiceDifferentiable,
+    training_data: DataLoader,
+    b: torch.Tensor,
+    lam: float = 0,
+    progress: bool = True,
+) -> torch.Tensor:
+    """Given a model and training data, it finds x s.t. $Hx = b$, with $H$ being
+    the model hessian.
 
+    :param model: A model wrapped in the TwiceDifferentiable interface.
+    :param training_data: A DataLoader containing the training data.
+    :param b: a vector or matrix
+    :param lam: regularization of the hessian
+    :param progress: If True, display progress bars.
 
-def as_tensor(a: Any, warn=True, **kwargs):
+    :return: An array that solves the inverse problem,
+        i.e. it returns $x$ such that $Hx = b$
+    """
+    all_x, all_y = [], []
+    for x, y in training_data:
+        all_x.append(x)
```

Comment:
Isn't there a risk to run out of memory doing it this way?

Reply:
yes. That is why the direct method can be used only with tiny models and training sets and typically approximate methods are used

Reply:
There is one alternative, which is to keep one grad per batch, like it is done for cg. Maybe we can discuss what makes more sense.